JavaScript development/Frameworks/React.md

```diff
@@ -0,0 +1,598 @@
+    ).isRequired,
+  };
+
+  static defaultProps = {
+    heading: 'Default Heading',
+  };
+
+  // ...
+
+  render() {
+    const {name} = this.props;
+
+    return (
+      <div>{name}</div>
+    );
+  }
+}
+```
+
+In stateless components, use PropTypes for type checking but use native JS for setting default values.
+
+``` jsx
+function ComponentName({id, name = 'John Doe'}) {
+  // ...
+};
+
+ComponentName.propTypes = {
+  id: PropTypes.number.isRequired,
+  name: PropTypes.string
+}
+```
+
+### Using props
+
+If you are using more than one property in a method, preffer accessing them with **object destructuring**.
+
+``` jsx
+export class ComponentName extends React.Component {
+  // ...
+  render() {
+    const {id, name, onClickHandler} = this.props;
+    // ...
+  }
+}
+```
+
+### Importing React
+``` jsx
+import React from 'react';
+
+export class ComponentName extends React.Component {
+  // ...
+}
+```
+
+### Exporting Components
+- each React component file should have a single exported component
+- don't use default exports, always use named exports instead
+ 
+Exporting class components:
+``` jsx
+// bad
+export default class ComponentName extends React.Component {
+  // ...
+}
+
+// good
+export class ComponentName extends React.Component {
+  // ...
+}
+```
+
+Exporting class components when you need to declare static members afterwards:
+``` jsx
+// exporting when static members are not supported by the transpiler:
+class ComponentName extends React.Component {
+  // ...
+}
+
+ComponentName.propTypes = {
+  // ..
+};
+
+export {ComponentName};
+```
+
+Exporting functional components:
+``` jsx
+// stateless components with no prop validation
+export function FunctionalComponent() {
+  // ...
+};
+
+
+// stateless components with static prop validation
+function FunctionalComponent(/* props */) {
+  // ...
+}
+
+FunctionalComponent.propTypes = {
+  // ...
+};
+
+export {FunctionalComponent};
+```
+
+### Rendering Arrays of Data
+
+For iterating arrays:
+- use **map** method to render items
+- you should have a unique key for each element
+- **don't use array's index as a key**, unless you have no choice (no unique property in each element of the array)
+
+``` jsx
+render() {
+  // ...
+  return (
+    <ul>
+      {
+        booksArray.map((book) => (
+          <li key={book.id}>
+            <a href={book.href}>{book.title}</a>
+          </li> 
+        )); 
+      }
+    </ul>
+  );
+}
+```
+
+### Binding
+- don't bind functions inline or in a class constructor
+- use [`developit/decko`](https://github.com/developit/decko) lib's *@bind* decorator for binding
+
+``` jsx
+// bad
+export class ComponentName extends React.Component {
+  onClickHandler(e) {/**/}
+
+  render() {
+    return (
+      <ChildComponent onClick={this.onClickHandler.bind(this)} />
+    );
+  }
+}
+
+// bad
+export class ComponentName extends React.Component {
+  constructor(args) {
+    super(args);
+    this.onClickHandler = this.onClickHandler.bind(this);
```

Comment:
Wouldn't say this is necessary bad. This is the recommended way if you don't have decorators available (they're not a standard feature, we're adding them because of mobx)

Reply:
If the project is not using mobx it can still use `bind` decorator so in this case, I would consider this an anti-pattern.

Reply:
But what if we don't have decorators available?